pandora-server-directory/src/database/mongoDb.ts

```diff
@@ -593,6 +608,79 @@ export default class MongoDatabase implements PandoraDatabase {
 
 	private async doManualMigrations(): Promise<void> {
 		// Add manual migrations here
+
+		//#region Migrate character self info from accounts to characters
+		const charactersToMigrate = new Map<CharacterId, { account: AccountId; character: DatabaseCharacterSelfInfo; }>();
+
+		// Gather data about characters
+		await accountCollection.doManualMigration(this._client, this._db, {
+			oldSchema: DatabaseAccountWithSecureSchema.extend({
+				characters: DatabaseCharacterSelfInfoSchema
+					.omit({ currentSpace: true })
+					.and(z.object({ currentRoom: SpaceIdSchema.nullable().optional() }))
+					.array()
+					.optional(),
+			}),
+			migrate: async ({ oldStream }) => {
+				for await (const account of oldStream) {
+					if (Array.isArray(account?.characters)) {
+						for (const character of account.characters) {
+							Assert(!charactersToMigrate.has(character.id));
+
+							charactersToMigrate.set(character.id, {
+								account: account.id,
+								character: {
+									id: character.id,
+									name: character.name,
+									preview: character.preview,
+									currentSpace: character.currentRoom ?? null,
+									inCreation: character.inCreation,
+								},
+							});
+						}
+					}
+				}
+			},
+		});
+
+		// Apply the data to the characters
+		await characterCollection.doManualMigration(this._client, this._db, {
+			oldSchema: CharacterDataSchema.pick({ accountId: true, name: true, inCreation: true }).and(z.object({ id: z.number().int() })),
+			migrate: async ({ oldCollection, oldStream, migrationLogger }) => {
+				for await (const character of oldStream) {
+					if (character == null)
+						continue;
+
+					const migrationInfo = charactersToMigrate.get(CharacterId(character.id));
+					if (migrationInfo == null)
+						continue;
+
+					Assert(character.accountId === migrationInfo.account);
+					Assert(CharacterId(character.id) === migrationInfo.character.id);
+					Assert(character.name === migrationInfo.character.name);
+					Assert(character.inCreation === migrationInfo.character.inCreation);
+
+					migrationLogger.verbose(`Migrating character ${migrationInfo.account}/${migrationInfo.character.id}`);
+
+					const { matchedCount } = await oldCollection.updateOne(
+						{ id: character.id },
+						{
+							$set: {
+								currentSpace: migrationInfo.character.currentSpace,
+								preview: migrationInfo.character.preview,
+							},
+						},
+					);
+					Assert(matchedCount === 1);
+
+					charactersToMigrate.delete(migrationInfo.character.id);
+				}
+			},
+		});
+
+		// The character array from the account will be deleted during automatic migration
```

Comment:
there should be a log with total number of migrated acc/chars and we could also log that `charactersToMigrate.size` is 0 at the end